### PR TITLE
Add tun network interface implementation

### PR DESCRIPTION
**Pull request by francescolavra**
The tun interface driver is implemented as a klib.
In order to allow adding a new type of special files and implementing blocking I/O on this file, I/O operations on special files have been refactored using custom fdesc closures, instead of relying on the fdesc closures used for regular files.

### CODE REVIEW

**Review by sanderssj — COMMENTED** [review]
LGTM

File: klib/tun.c [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+closure_function(1, 6, sysreturn, tun_write,
+                 tun, tun,
+                 void *, src, u64, len, u64, offset, thread, t, boolean, bh, io_completion, completion)
+{
+    tun tun = bound(tun);
+    if (!tun->netif)
+        return io_complete(completion, t, -EBADFD);
+    if (!(tun->flags & IFF_NO_PI)) {
+        if (len < sizeof(struct tun_pi))
+            return io_complete(completion, t, -EINVAL);
```
Comment (wjhun) [inline klib/tun.c:151]: Should len == sizeof(struct tun_pi) just return io_complete(completion, t, len) here?

**Review by wjhun — APPROVED** [review]
Everything looks good...just one question about pi write with zero len payload

File: klib/tun.c (hunk 1 above)
Comment (francescolavra) [reply klib/tun.c:154]: Indeed, calling pbuf_alloc() with len=0 is wrong. Fixed.
If the IFF_NO_PI flag is not present, a write() call with len == sizeof(struct tun_pi) returns sizeof(struct tun_pi), while if the IFF_NO_PI flag is present, a write() call with len == 0 returns -EINVAL. This matches the behavior of the Linux kernel.

**Pull request merged**